File: flip.cpp
#include <cstdlib>
#include <vector>
#include <iostream>
#include <string>

using namespace std;

int compare(int a, int b)
{
	int f[3]={a,b,b};
	
	int k=a-b;
	while((k!=1 && k!=-1)&& k!=0)
	{
		k/=2;
	}
//	cout<<k<<", "<<f[-1]<<", "<<f[k]<<endl;
	return f[k+1];
}

vector<int> flip(vector<int> arr)
{
	int restore=arr[0];
	vector<int> newarr=arr;
	vector<int>::iterator itr;
	int count=0;

	for(itr=newarr.begin()+1;itr!=newarr.end();)
	{
		if(restore== *(itr))
		{
			newarr.erase(itr);
		}
		else
		{
			restore= *(itr);
			++itr;
		}
	}
	for(itr=newarr.begin();itr!=newarr.end();++itr)
	{
		count+= *itr;
	}
//	cout<<"for flip : "<<compare(count,newarr.size()-count);
//	cout<<count<<", "<<newarr.size()<<endl;
	cout<<compare(count,newarr.size()-count);
	return newarr;
}

vector< vector<int> > init()
{
	vector< vector<int> > output;
	vector<int> col;

	col.push_back(0);
	col.push_back(0);
	output.push_back(col);
	col.clear();
	
	col.push_back(0);
	col.push_back(1);
	output.push_back(col);
	col.clear();
	
	col.push_back(1);
	col.push_back(0);
	output.push_back(col);
	col.clear();
	
	col.push_back(1);
	col.push_back(1);
	output.push_back(col);
	col.clear();
	
	return output;
}

int flip02(vector<int> input)
{
	vector< vector<int> > mat=init();
	vector<int> restore;
	vector<int> pattern;
	vector<int> inpat;
	vector<int> various;
	vector<int>::iterator itr;
	vector< vector<int> >::iterator row;
	int value=0;
	int idx=0;
	int len=input.size();
	int i=1;
	int k=0;
	
	for(k=0;k<4;++k)
	{
		inpat.push_back(k);
		various.push_back(0);
	}
	
	for(itr=input.begin();itr!=input.end();++itr)
	{
		value += *itr;
	}
	
	for(i=1,itr=input.begin(),row=mat.begin();i<len;++i)
	{
		idx=0;
		restore.clear();
		restore.push_back(*(itr+i-1));
		restore.push_back(*(itr+i));
		while(*(row+idx) != restore)
		{
			++idx;
		}
		pattern.push_back(*(inpat.begin()+idx));
	}
	
	for(itr=pattern.begin();itr!=pattern.end();++itr)
	{
		++*(various.begin()+*itr);
	}
	
	itr=various.begin();
	
	return *(itr+1+*input.begin());
}

void fliptest(void)
{
	vector<int> arr;
	vector<int> newarr;
	vector<int>::iterator itr;
	vector<char>::iterator c_itr;
	int idx=0;
	string str;
	
	cin>>str;

//	for(c_itr=str.begin();c_itr!=str.end();++itr)
	for(idx=0;idx<str.size();++idx)
	{
		arr.push_back(static_cast<int>(str[idx]-48));
	}
/*	
	for(idx=0;idx<=100;++idx)
	{
		arr.push_back(rand()%2);
	}*/
	
/*	
	for(itr=arr.begin();itr!=arr.end();++itr)
	{
		cout<<*itr<<" | ";
	}
	cout<<endl;*/
	newarr=flip(arr);
/*	cout<<"to flip : "<<endl;
	for(itr=newarr.begin();itr!=newarr.end();++itr)
	{
		cout<<*itr<<" | ";
	}*/
}

void fliptest02(void)
{
	vector<int> arr;
	vector<int> newarr;
	vector<int>::iterator itr;
	vector<char>::iterator c_itr;
	int idx=0;
	string str;
	
	cin>>str;

	for(idx=0;idx<str.size();++idx)
	{
		arr.push_back(static_cast<int>( str[idx]&1 ) );
	}

	newarr=flip(arr);
}

void fliptest03(void)
{
	vector<int> arr;
	int idx=0;
	string str;
	
	cin>>str;

	for(idx=0;idx<str.size();++idx)
	{
		arr.push_back(static_cast<int>( str[idx]&1 ) );
	}
	
	cout<<flip02(arr)<<endl;
}

int main(void)
{
	fliptest03();
	
	return 0;
}

File: flip.py
# -*- coding: utf-8 -*-
"""
Created on Sat Dec  3 01:30:44 2022

@author: NOPE
"""
import copy
import random

def compare( a,b ):
    f = [a,b,a]
    k=a-b
    while((k!=1 and k!=-1)and k!=0):
        k=k//2
    return f[k]

def flip(arr):
    size=len(arr)
    restore=0
    result=0
    for i in range(arr):
        restore+=i
    result=compare(restore,size-restore)
    return result

def flip01(arr):
    idx=0
    wheredf=[]
    wheresm=[]
    where=[]
    for i in range(arr):
        if(idx!=idx+i):
            wheresm.append(copy.deepcopy(where))
            where.clear()
            where.append(idx)
        elif(idx==idx+i):
            wheredf.append(copy.deepcopy(where))
            where.clear()
            where.append(idx)
        idx+=1
    
    return 0

def flip02(arr):
    idx=0
    dfsize=0
    smsize=0
    wheredf=[]
    wheresm=[]
    where_d=[]
    where_s=[]
    for i in arr:
        if(idx!=idx+i):
            wheresm.append(copy.deepcopy(where_s))
            smsize+=1
            where_s.clear()
            where_d.append(idx)
        elif(idx==idx+i):
            wheredf.append(copy.deepcopy(where_d))
            dfsize+=1
            where_d.clear()
            where_s.append(idx)
        idx+=1
    print(wheredf)
    print(wheresm)
    return compare(smsize,dfsize)

def flip03(arr):
    restore=arr[0]
    newarr=[arr[0]]
    count=0
    for i in arr:
        if(restore!=i):
            restore=i
            newarr.append(restore)
    
    for j in newarr:
        count+=j
    print(newarr)
    return compare(count,len(newarr)-count)

def flip04(arr):
    restore=arr[0]
    newarr=copy.deepcopy(arr)
    count=0
    for i in arr:
        if(restore==newarr[count]):
            newarr.remove(newarr[count])
            count-=1
        else:
            restore=i
        print(i,newarr,count)
        count+=1
    count=0
    for j in newarr:
        count+=j
    print(newarr)
    return compare(count,len(newarr)-count)

def flip04A(arr):
    restore=arr[0]
    newarr=copy.deepcopy(arr)
    ip=1
    # print(newarr[1:])
    for i in arr:
        if(i==newarr[ip]):
            newarr.remove(newarr[ip])
        else:
            restore=i
            ip+=1
        print(i,newarr,ip)
        
    count=0
    for j in newarr:
        count+=j
    print(newarr)
    return compare(count,len(newarr)-count)

def flip04A01(arr):
    restore=arr[0]
    newarr=copy.deepcopy(arr)
    idx=1
    # print(newarr[1:])
    for i in arr[1:]:
        if(restore==newarr[idx]):
            newarr.pop(idx)
        else:
            restore=newarr[idx]
            idx+=1
        # print(restore,newarr,idx)
        
    count=0
    for j in newarr:
        count+=j
    # print(newarr)
    print(count, len(newarr))
    return compare(count,len(newarr)-count)

def flip05(arr):
    mat=[[0,0],[0,1],[1,0],[1,1]]
    restore=[]
    pattern=[]
    value=0
    variousU=0
    variousD=0
    idx=0
    size=len(arr)
    for j in arr:
        value += j
    
    for i in range(1,size):
        idx=0
        restore.clear()
        restore.append(arr[i-1])
        restore.append(arr[i])
        while(restore != mat[idx]):
            idx += 1
        if(idx == 1):
            pattern.append('u')
        elif(idx == 2):
            pattern.append('d')
    
    for k in pattern:
        if(k == 'u'):
            variousU += 1
        else:
            variousD += 1
    
    if(value >= size-value):
        return variousU
    else:
        return variousD

def flip05A(arr):
    mat=[[0,0],[0,1],[1,0],[1,1]]
    restore=[]
    pattern=[]
    value=0
    variousU=0
    variousD=0
    idx=0
    size=len(arr)
    for j in arr:
        value += j
    
    for i in range(1,size):
        idx=0
        restore.clear()
        restore.append(arr[i-1])
        restore.append(arr[i])
        while(restore != mat[idx]):
            idx += 1
        if(idx == 1):
            pattern.append('u')
        elif(idx == 2):
            pattern.append('d')
    
    for k in pattern:
        if(k == 'u'):
            variousU += 1
        else:
            variousD += 1
    
    if(value >= size-value):
        if(variousU == 0):
            return variousD
        return variousU
    else:
        if(variousD == 0):
            return variousU
        return variousD

def flip05A01(arr):
    mat=[[0,0],[0,1],[1,0],[1,1]]
    restore=[]
    pattern=[]
    value=0
    variousU=0
    variousD=0
    idx=0
    size=len(arr)
    for j in arr:
        value += j
    
    for i in range(1,size):
        idx=0
        restore.clear()
        restore.append(arr[i-1])
        restore.append(arr[i])
        while(restore != mat[idx]):
            idx += 1
        if(idx == 1):
            pattern.append('u')
        elif(idx == 2):
            pattern.append('d')
    
    for k in pattern:
        if(k == 'u'):
            variousU += 1
        else:
            variousD += 1
    
    if(value >= size-value):
        if(variousU == 0):
            return variousD
        if(arr[0]==1):
            variousU += 1
        return variousU
    else:
        if(variousD == 0):
            return variousU
        return variousD

def flip05A02(arr):
    mat=[[0,0],[0,1],[1,0],[1,1]]
    restore=[]
    pattern=[]
    value=0
    variousU=0
    variousD=0
    idx=0
    size=len(arr)
    for j in arr:
        value += j
    
    for i in range(1,size):
        idx=0
        restore.clear()
        restore.append(arr[i-1])
        restore.append(arr[i])
        while(restore != mat[idx]):
            idx += 1
        if(idx == 1):
            pattern.append('u')
        if(idx == 2):
            pattern.append('d')
    
    for k in pattern:
        if(k == 'u'):
            variousU += 1
        elif(k == 'd'):
            variousD += 1
    
    if(arr[0] == 1):
        return variousD
    elif(arr[0] == 0):
        return variousU
        
def flip05A03(arr):
    mat=[[0,0],[0,1],[1,0],[1,1]]
    restore=[]
    pattern=[]
    inpat=[0,1,2,3]
    value=0
    various=[0,0,0,0]
    idx=0
    size=len(arr)
    for j in arr:
        value += j
    
    for i in range(1,size):
        idx=0
        restore.clear()
        restore.append(arr[i-1])
        restore.append(arr[i])
        while(restore != mat[idx]):
            idx += 1
        pattern.append(inpat[idx])

    for k in pattern:
        various[k] += 1
    
    return various[1+arr[0]]

def fliptest():
    # print(flip04A01([0,0,1,1,1,0,0,0,0,1,1,1,1,1,0]))
    arr=[0,0,0,1,1,0,0]
    print(flip04A01(arr))

def fliptest01():
    arr=[]
    instr=input()
    idx=len(instr)
    for i in range(idx):
        arr.append(ord(instr[i])-48)
    
    print(flip04A01(arr))

def fliptest02():
    
    instr=input()
    N=len(instr)
    arr=[0]*(N)
    for k in range(N):
        arr[k]=ord(instr[k])&1
    
    print(flip04A01(arr))

def fliptest03():
    intable=[1,0,0,1,1,0,1,0,1,1,1,0]
    output=0
    output=flip05A03(intable)
    print(output)
    

def main():
    fliptest03()

if(__name__=="__main__"):
    main()